crypto/s2n_rsa_pss.c

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <openssl/evp.h>
+#include <stdint.h>
+
+#include "error/s2n_errno.h"
+
+#include "stuffer/s2n_stuffer.h"
+
+#include "crypto/s2n_hash.h"
+#include "crypto/s2n_openssl.h"
+#include "crypto/s2n_rsa.h"
+#include "crypto/s2n_pkey.h"
+
+#include "utils/s2n_blob.h"
+#include "utils/s2n_random.h"
+#include "utils/s2n_safety.h"
+#include "utils/s2n_blob.h"
+
+#define RSA_PSS_SIGN_VERIFY_RANDOM_BLOB_SIZE    32
+#define RSA_PSS_SIGN_VERIFY_SIGNATURE_SIZE      256
+
+#if RSA_PSS_SUPPORTED
+
+const EVP_MD* s2n_hash_alg_to_evp_alg(s2n_hash_algorithm alg) {
+    switch (alg) {
+        case S2N_HASH_MD5_SHA1:
+            return EVP_md5_sha1();
+        case S2N_HASH_SHA1:
+            return EVP_sha1();
+        case S2N_HASH_SHA224:
+            return EVP_sha224();
+        case S2N_HASH_SHA256:
+            return EVP_sha256();
+        case S2N_HASH_SHA384:
+            return EVP_sha384();
+        case S2N_HASH_SHA512:
+            return EVP_sha512();
+        default:
+            return NULL;
+    }
+}
+
+static int s2n_rsa_pss_size(const struct s2n_pkey *key)
+{
+    notnull_check(key);
+
+    /* For more info, see: https://www.openssl.org/docs/man1.1.0/man3/EVP_PKEY_size.html */
+    return EVP_PKEY_size(key->key.rsa_pss_key.pkey);
+}
+
+/* On some versions of OpenSSL, "EVP_PKEY_CTX_set_signature_md()" is just a macro that casts digest_alg to "void*",
+ * which fails to compile when the "-Werror=cast-qual" compiler flag is enabled. So we work around this OpenSSL
+ * issue by creating a local non-const duplicate of EVP_MD pointer, calling OpenSSL with the duplicate, then freeing the
+ * duplicate since our compiler won't let us pass a const pointer to an API that doesn't guarantee to not modify it. */
+static int s2n_evp_ctx_set_signature_digest(EVP_PKEY_CTX *ctx, const EVP_MD* const_digest_alg, int s2n_err) {
+    notnull_check(ctx);
+    notnull_check(const_digest_alg);
+
+    EVP_MD* digest_alg = EVP_MD_meth_dup(const_digest_alg);
+    notnull_check(digest_alg);
+
+    int rc = EVP_PKEY_CTX_set_signature_md(ctx, digest_alg);
+    EVP_MD_meth_free(digest_alg);
+    S2N_ERROR_IF(rc <= 0, s2n_err);
+
+    return 0;
+}
+
+static int s2n_rsa_pss_sign(const struct s2n_pkey *priv, struct s2n_hash_state *digest, struct s2n_blob *signature_out)
```

Comment:
I'm not sure if this method actually does what you think. The problem is that the hash function is used three different times in PSS signing and you might only be setting two of them.

1. Hash of raw message to product `mHash`. (I _think_ that the output of this is what is being passed in via `digest`.)
2. Hash of  `M' = padding1 || mHash || salt` to product H. (I _think_ this is correctly being done because of your call to `EVP_PKEY_CTX_set_signature_md`, but I'm not sure.)
3. Hash function used internally by MGF1. (This is not being set and may default to SHA-1 as that is from the base MGF1 specification. The documentation doesn't make it clear what OpenSSL will do in this case. You should call `EVP_PKEY_CTX_set_rsa_mgf1_md` to make it explicit what should be done.)

Reply:
Fixed

Reply:
Fixed